### ISSUE TITLE
REST API

### ISSUE THREAD

**Issue by joshdanielson** [loconomics/loconomics#234]
@IagoSRL commented on [Tue Jul 15 2014](https://github.com/joshdanielson/loconomics-internal/issues/530)

Tracking implementation of the REST API.
More issues for specific REST APIs will follow.
### Implementation Status
- [x] Decide and implement final URL for the API and language and version management.
  - Initial one: /en-US/rest/
  - **Final one**: /api/v1/en-US/  Code adapted to this already, with the locale part being dynamic.
- [x] Events: #579 
- [x] Event Types
- [x] Availability: read user calendar availability with granularity: 
  - weekly, 
  - monthly, 
  - monthly-schedule (editable too), 
  - weekly-schedule (editable too).
- [x] Availability types.
- [x] Login: implementation based on custom headers to pass the authentication; includes cookie for other website uses but can get removed in a future with its Logout counter-part.
  - [x] Facebook login implemented, not tested from devices.
- [X] Signup
  - [x] Facebook signup implemented, not tested from devices.
- [x] Upcoming bookings: updated to final app mockups requirements.
- [x] Bookings: #580 read bookings and edit freelancer-bookings.
- [x] Pricing: #581
- [x] JobTitles: get information about a job title and what pricing types uses.
  - [x] Search option for Autocomplete components /job-titles/autocomplete
  - [x] Allow users to add a job title if don't exists #650 
- [x] PricingTypes: get complete information of pricing types in order to implement forms and messaging logic.
- [x] UserJobProfile: get the list of job titles for the logged user; update the basic info (table [userprofilepositions]: intro, cancellation policy, instantBooking preference) and add a job title to the profile; request job title deactivation and reactivation, latest is constrained to fulfill the required alerts for an active profile, same back-end code as website.
- [x] Clients: #646
- [x] Locations #677
- [x] Inbox / Messaging #678
- [x] Schedule preferences: #683
- [x] Calendar Syncing: #684 _just provide URLs for export and import ical_
- [x] User Profile / contact information: #685:
  - [x] /profile: account and contact info
  - [x] /marketplace-profile: marketplace specific profile info
- [x] Privacy Settings: #686
- [x] Feedback and contact forms: VOCFeedback table and support email, see #688

**DUDE**
- [ ] REST API support to create Client Location by Freelancer: #725 

**Payment Processing specifics**
- [x] Payment processing: #689
  - [x] Merchant account set-up (with Braintree)
  - [x] Fetch identification info (payout method, latest account numbers, status).
- [ ] Transaction history #690
  - [ ] Past transactions by dates (_transaction report_).
  - [ ] Pending transactions
- [ ] Payment processing enabled in Freelancer Bookings #676

**Co-op/Membership specifics**
- [ ] User fees: #687
  - [ ] Next payment info
  - [ ] Payment plan change
  - [ ] Payment method, store in Braintree
  - [ ] Payment history

**Marketplace specifics**
- [x] Public bio #691
- [x] Profile picture #692
- [ ] Languages #693 _review how relate with service attributes_
- [ ] Verifications #694
  - [x] Fetch user verifications
  - [ ] Request to perform a verification (different types: email, facebook)
- [x] Education #695
- [ ] Background checks #696
- [x] Brief description, AKA 'Job title intro', is part of the UserJobProfile API. Is in 'Overview of services' in the UI mockups.
- [x] Service attributes: #697. is in 'Overview of services' in UI mockups
  - [x] Retrieve and save selection
  - [x] Autocomplete/search of available service attributes for the user job title
  - [x] Create new attribute if not exists.
- [x] Professional certifications/licenses #698 
  - [x] read user certifications/licenses
  - [x] request verification of certification/license by uploading photo
- [x] Images of my work #699
- [x] Cancellation policy, scheduling confirmation preference and job title intro are included in the UserJobProfile, no need for more specific API even if they are different sections in the UI.
- [ ] Cancellation policy types: #700 Retrieve information about available types.

**Next, candidate for version 1.1**
- [x] Availability, weekly schedule: **change** rather than time slots, the API must provide the list of daily time-ranges. It's less data and how database really save the data, while the App UI is closer to this (right now limited to 1 time-range per day, but will be enhanced to allow multiple time ranges, so matches the database and allow any set-up) and the previous Website UI (the weekly calendar editable per slot) will be replaced by the enhanced App UI.
- [ ] Freelancer Edit Booking Price: to allow freelancers to change the booking price, per item/pricing, replacing the preset one and recalculating total. #680 
- [ ] Performance: #682
- [ ] Client history: #681 _(bookings queries: upcoming bookings, past, need schedule; messages filtering per client)_

**Future/revisions**
- [ ] Housekeeper specific pricing features
- [ ] Babysitter specific pricing features
- [ ] Pricing variables, how this works with housekeeper and babysitter pricing.
- [ ] FAQs: initially loaded in code in the app, may require to be in the API to allow remote updates without require to publish a new app version. Needs strong caching, pings to notify updates.
- [ ] Client profile specifics

**Client booking**
- [x] /me  Created this new entry point for all APIs that manage data of the currently logged user, not allowing to specify the userID on the requests for the data being returned, making clean and secure separation of API purposes and requiring some different entry points and scheme for similar tables.
  - [X] /me/client-booking Allow a client to book a freelancer. It has different rules than a freelancer creating a booking (/freelancer-booking).
  - [x] /me/payment-methods Allow a client to retrieve partial information of any saved method/cards from previous booking payments, data comes from Braintree. Is used to simplify the payment process on client booking, like in website bookings (same code, but carefully exposing the data). It has a Braintree ID/token that can be used for later payment, but full details cannot be fetched with this (Braintree does not allow that).
- [x] /users/{userID}  Created this new entry point for all APIs that ask to retrieve information about another user (not the logged user), so the retrieved information is made **public** by the owner user or available only to users that have connected with it previously.
  - [x] /users/{userID} The index entry point gets a summary and grouped results from several specific _users/ API_ so is just one request to get most used user public info, like when starting a client-booking. It includes right now: profile, rating, verificationsSummary, jobProfile.
  - [x] /users/{userID/freelancer-pricing/{jobTitleID} Public pricing info per jobTitle, to display on a client booking. Must be queried per jobTitleID like in the example  /users/141/freelancer-pricing/106 -> gets the user 141 pricings for the job title 106
  - [x] /users/{userID}/job-profile Fetch user public job profile, including its active job titles only (a private, "/me/" equivalent will show even inactive titles because the owner can choose to enable them or manage them, but never disclosed publicly in this API).
  - [x] /users/{userID}/profile Fetch user public profile information. Basic info as name (firstName, lastName,..), publicBio, contact data (email, phone), isFreelancer flag. When freelancer, other fields are included as businessName, websiteUrl (is contact data), profileUrlSlug. The contact data is included only if the user requesting the data (logged user) has a formal relationship with the requested user (when both users are _connected_, record in ProviderCustomer table).
  - [x] /users/{userID}/ratings Fetch user ratings, global user rating on this URL, and rating as customer, as freelancer (average from job titles) and specific job title rating in following entry points:
    - [x] /users/{userID}/ratings/client
    - [x] /users/{userID}/ratings/freelancer
    - [x] /users/{userID}/ratings/{jobTitleID}  like /users/141/ratings/106
  - [x] /users/{userID}/verifications-summary Fetch user verifications summary (total of verifications, and total verifications by groups like verifications, reviews and licenses), global user verifications on this URL, and as customer, as freelancer (average from job titles) and specific job title rating in following entry points:
    - [x] /users/{userID}/verifications-summary/client
    - [x] /users/{userID}/verifications-summary/freelancer
    - [x] /users/{userID}/verifications-summary/{jobTitleID}  like /users/141/verifications-summary/106
  - [x] /users/{userID}/service-addresses Fetch public freelancer service addresses per job title.
  - [x] /users/{userID}/availability Publicly allow to query a provider availability. The API was created before as /availability but following a legacy approach used by the website widgets (R6 dashboard), adjusted here to follow the /user/{userID}/ pattern and the entry points and results format:
    - [x] /users/{userID}/availability/times Retrieve specific available time slots per date (used on detailed date view, called weekly previously with a different set of default params).
    - [x] /users/{userID}/availability/dates Retrieve if a date has some availability or is completely unavailable (used on monthly views, called monthly previously with a different set of default params)

---

@IagoSRL commented on [Thu Jul 17 2014](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-49345371)

Hi @derekrazo,

I uploaded initial work on the REST API to _dev_, you can test it from the next testing page: http://dev.loconomics.com/en-US/Tests/TestREST

Its just like a _thin client_ with jQuery to do quick REST requests on the site.
I included a list of current REST API URLs and data examples (as you can see in the page source code, its just a JS object with each example configuration per preferred http-method, but not limited to it, built as an ul-list on the page bottom).
I created a fake API /Test page just to try how to implement REST with asp.net, building the common parts and a boilerplate code to reuse when implementing the actual APIs. With this, I solved already some common problems so next could be built quickly (I hope!). It has examples of returning error messages and validation/input errors (as an object with an 'error strings array' per property).

There is too some actual APIs already:
- **upcoming-announcements**: you have example links in the page and some helping comments; its very easy, and only needs GET support. It was easy to put in place once I get the boilerplate from _/Test_.
- **calendar, get-availability (workhours, monthly, monthly-schedule, weekly)**: this is _old_ really, but I put examples on the page about how to use them. This is already being used by current calendar widgets in the _live_ site. You can see that the results are wrapped in a common scheme {Code,Result}. It can be unwrapped and just return the content of _Result_ since the _Code_ is meaningless and is better to return _http error codes_ and the _errorMessage_ property.
  We were using this JSON scheme around the site along with some predefined codes that trigger specific javascript functions after save a from via AJAX (a popup with the text at _result_, a client side redirect or page refresh, etc.), but this kind of logic is not need as part of a RESTful API so may be better to avoid them, right?

There are some other APIs that return JSON for several places, just a few, like one that returns data for a jquery-ui autocomplete, or server side validation of postal code.

Please, take a look and comment about any convention or best practice you think we can follow to build the complete RESTful API in a convenient way, any change you may desire to the current implemented ones.

cheers

---

@derekrazo commented on [Mon Jul 21 2014](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-49686777)

Hi Iago!

My experience building REST API's from scratch is far more limited that my
experience using them :).

Happy to defer to your judgement- or if you would like some critical peer
review I can bring in someone from my network to talk to you.

I'm going to stay focused on mobile design, user interviews & testing and
more "high level" dev stuff for now.

Best,

Derek

On Thu, Jul 17, 2014 at 11:29 AM, IagoSRL notifications@github.com wrote:

> Hi @derekrazo https://github.com/derekrazo,
> 
> I uploaded initial work on the REST API to _dev_, you can test it from
> the next testing page: http://dev.loconomics.com/en-US/Tests/TestREST
> 
> Its just like a _thin client_ with jQuery to do quick REST requests on
> the site.
> I included a list of current REST API URLs and data examples (as you can
> see in the page source code, its just a JS object with each example
> configuration per preferred http-method, but not limited to it, built as an
> ul-list on the page bottom).
> I created a fake API /Test page just to try how to implement REST with
> asp.net, building the common parts and a boilerplate code to reuse when
> implementing the actual APIs. With this, I solved already some common
> problems so next could be built quickly (I hope!). It has examples of
> returning error messages and validation/input errors (as an object with an
> 'error strings array' per property).
> 
> There is too some actual APIs already:
> - _upcoming-announcements_: you have example links in the page and
>   some helping comments; its very easy, and only needs GET support. It was
>   easy to put in place once I get the boilerplate from _/Test_.
> - _calendar, get-availability (workhours, monthly, monthly-schedule,
>   weekly)_: this is _old_ really, but I put examples on the page about
>   how to use them. This is already being used by current calendar widgets in
>   the _live_ site. You can see that the results are wrapped in a common
>   scheme {Code,Result}. It can be unwrapped and just return the content of
>   _Result_ since the _Code_ is meaningless and is better to return _http
>   error codes_ and the _errorMessage_ property. We were using this JSON
>   scheme around the site along with some predefined codes that trigger
>   specific javascript functions after save a from via AJAX (a popup with the
>   text at _result_, a client side redirect or page refresh, etc.), but
>   this kind of logic is not need as part of a RESTful API so may be better to
>   avoid them, right?
> 
> There are some other APIs that return JSON for several places, just a few,
> like one that returns data for a jquery-ui autocomplete, or server side
> validation of postal code.
> 
> Please, take a look and comment about any convention or best practice you
> think we can follow to build the complete RESTful API in a convenient way,
> any change you may desire to the current implemented ones.
> 
> cheers
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/dani0198/Loconomics/issues/530#issuecomment-49345371.

---

@IagoSRL commented on [Fri Feb 27 2015](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-49721095)

Hi Derek,

I understand you want to focus on what is priority now, so take your time answering this.

I just want your opinion **as a consumer** of the API, just the **interface**. Some of this decisions may be quick to change in an implemented API (before its being actively used) so may not be urgent.
And as commented already, the time consuming work is implement the underlying features, not the interface questions commented here.

Some doubts/examples: **_edited 2015-02-27: finally, the chosen ones are marked, just for documentation purposes**_
- Success HTTP codes preference:
  - All success requests must return _200_
  - [**chosen**] Or the more specific, as _201 Created_ for POST (inserted new record)
  - and the specific _204 No Content_ for DELETE (just deleted, no data returned). [**use this if makes more sense on some cases, but by default will be 200 on DELETE returning a copy of the deleted data**]
- Data returned for post (insert) / put (update) / delete operations:
  - nothing must be returned
  - [**chosen**] the inserted or updated row must be returned, with any auto-incremental ID assigned, or even any server calculated data (createdDate, modifiedDate... or any special for each case).
  - for Delete:
    - nothing
    - [**chosen**] or a copy of the deleted one
- PUT operations must:
  - be able to performs as _update or insert_
  - [**chosen**] be strictly an update operation and return an error code, _404 Not Found_ when item doesn't exists.
- GET result-sets, a list/table of records:
  - [**chosen**] the returned data must be just an array of records (personally, I think next options are more extensible, but this one may work better with some front-end frameworks and to avoid some _glue code_)
  - the returned data will be an object with a generic property _rows_ (or another name) with the array of records
  - as previous, but with the entity name rather than _rows_ as property name.
  - some meta information included in other property objects, as total rows count, paging info or something else?

And similar stuff. I know that make everything follow a same scheme, when possible, can help to write more common and reusable code on front-end, and I want to avoid that every API ended with a different style, so if you find some of this things or others useful for you, let me know.

cheers,
Iago

---

@IagoSRL commented on [Mon Oct 27 2014](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-60590063)

### Status update of the REST API (from some comments on #578):
- Calendar availablity: **done**; options: weekly, monthly, monthly-schedule, workhours/weekly-schedule.
- Login: **done**, notes: its a cookie based solution, the same as in the website.
- Fetch upcoming-appointments: **done**.
- Events: **initial work, fetch done**; implemented right now the fetch of events with full details and basic filter options (date range and event types).

### Needed APIs for the mobile site to be implemented:
- Events: #579
  - Get details of a single event by ID
  - Get event(s) details simplified for the common set of used options.
  - Post/Put event, with simplified editing mode.
  - Block calendar dates; its a simplified way to create (POST) events per date for a dates range of busy events with the type 'weekly-schedule' (typeID:6).
- Bookings: #580
  - Making bookings (done by the freelancer).
  - Viewing bookings
  - Editing
- Pricing #581
  - Get list of pricings
  - Get pricing details
  - Create and update pricing
- Payment processing: **need more info**, @dani0198 What exactly we need now on the app (first version)? We don't need the payment process of customers on booking, because first version is for providers, and providers doesn't need to pay; maybe is just to allow providers to set-up their payment account?
- Adding new clients (depends on decisions to take on issue #570)

---

@IagoSRL commented on [Fri Feb 27 2015](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-76401909)

Updated first comment with task-lists and completed stuff, tracking progress. So many things, easy to easy to forget and lose yourself :-)

---

@joshdanielson commented on [Mon Mar 02 2015](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-76889089)

Just added and assigned issues to all of those that were missing.  We can chat tomorrow what more you'd like me to do beyond the mockup.  I know there will be some that need more info, but we can discuss that.  

---

@IagoSRL commented on [Thu Mar 05 2015](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-77403741)

Notes about URL scheme:
- Currently we have: /en-US/rest/
- Maybe better: /api/v1/en-US/
- I though at remove the language at first, being a user preference, parameter or header, but the database is designed to hold different data depending on language and country, that makes it not a trivial choice (like in other cases) else the provided data depends on that so needs to go in the URL. Some entry API may be offered locale agnostic if there are good reasons, but most data need it.
- Version or not? It may help to break things and make transitions. Against it: http://stackoverflow.com/questions/389169/best-practices-for-api-versioning. Supporting it: http://www.notmessenger.com/3-tenets-for-implementing-a-rest-api/. Summary of discussions and list of popular APIs: http://www.lexicalscope.com/blog/2012/03/12/how-are-rest-apis-versioned/

---

@joshdanielson commented on [Thu Mar 05 2015](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-77408588)

Great research!  I trust you :)  I think the /api/v1/en-US/ makes sense, but again you decide.  

---

@IagoSRL commented on [Mon Mar 09 2015](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-77917622)

Final URL scheme: /api/v1/en-US/, being the locale part dynamic (for future es-US, es-ES,.. any supported on the back-end), rather than a fixed folder/files structure.
Implemented updating all the APIs.

---

@IagoSRL commented on [Mon Mar 09 2015](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-77926817)

Josh, please confirm me next user profile information that will be available for edition in the API and the App:
- firstName
- middleIn **If Braintree doesn't offer this field, let's remove it.  Keep it if so**
- lastName
- secondLastName _we have it internally to be used conditionally on Spanish locale but not until now_  **Not need for US edition, but keep for future**
- nickName _Is this keep? I don't remember when appeared in database, unused I think_  **Let's keep it in the database but leave it out for this edition.  If we get feedback to add it, we will**
- publicBio
- genderID _Do we want the Gender still? I don't remember well the conversation about this_  **I've decided to just remove Gender altogether.  I suppose it doesn't hurt to keep it in the database just in case**
- mobilePhone _Renamed to just 'phone' in the API? so makes sense with the next new field_  **Yes, makes sense with the new field)**
- canReceiveSms _new_
- providerProfileUrl -> freelanceProfileUrl  **freelancerProfileUrl (you forgot the "r")**
- providerWebsiteUrl -> freelanceWebsiteUrl  **freelancerWebsiteUrl**
- birthMonthDay
- birthMonth

In the UI, the Profile and Website URLs I suppose goes on Marketplace profile on the section 'Your website' rather than 'Contact information', even it they are in the same table/API (can be split in the API if makes sense).  **Yes, it's good to have it separate**
PublicBio is edited on Marketplace Profile sub-section 'public bio'.
The birth month and day goes in Contact Information I suppose, right?  **Yes, that makes sense.**

There are other fields on the user profile table on database, some internal, other exposed as read-only, and other are user preferences/privacy settings that fits better in its own API. About this last, I think I named it wrong in the list (and then issue #686) and they are privacy settings, not user preferences.  **I updated #686 to Privacy Settings**

---

@joshdanielson commented on [Mon Mar 09 2015](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-77929358)

My answers are in **bold** above

---

@IagoSRL commented on [Mon Mar 09 2015](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-77930824)

Thanks!

Braintree does not has middleIn, so is out from the API.
In general, I will remove nothing from the database, just not exposed in the API.

Freelancer: actually, I put the 'r' in the code already and forgot here, ops!

URLs: fine, the separated API will be named? "Website", "Promote", "MarketplaceProfile" (and add  publicBio there)? This is quick to implement, just what name?

---

@IagoSRL commented on [Mon Mar 09 2015](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-77931775)

There is another field, read-only 'bookCode' that is used to build the book-me now button. Just auto-generated on first use, can go with that freelancer profile specialized API.

---

@joshdanielson commented on [Mon Mar 09 2015](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-77931916)

Okay, not sure I understand about the separate API...so you're saying a new API (so we will have two)--one with private information and another with public information?  If that is the case, then the public one could be called "Marketplace".  I think maybe I understand now...confirm :)

Also, there is one more field that's new that I forgot to add here:  businessName.  Corey, on the board, has his business named Tact Massage Therapy, for example.  That actually should help eliminate the need for a nickName field, too 

---

@IagoSRL commented on [Mon Mar 09 2015](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-77935375)

About the API, was more about to separate the user profile things (mostly on 'Contact Information') from the freelancer specific things (under Marketplace section on the UI) when using the API to access and update the data. It ends up that, as you pointed, the contact info is more sensitive, private in part, and the freelancer public, show in their marketplace profile when active.

Since everything goes in the same table on database right now, everything can go in the same API "/profile", the initial guess.

For example, for freelancer preferences related to the calendar, we have a [CalendarProviderAttributes] table on database, but its content is accessed from two APIs, already implemented following the naming in the App UI: /calendar-syncing and /scheduling-preferences

About API access rights, for now everything goes to be 'every user access its own information' with some exceptions; some APIs will require slight changes to add the possibility to fetch some/all of the data when accessed by another user (as customer asking for freelancer data, part of it obviously). Making implementation simple to avoid bugs on first App versions because of excessive initial complexity (there are already lot of things involved), but I keep thinking on the need to add that access later so current API makes sense.

I forgot about businessName one, adding it now.

---

@joshdanielson commented on [Mon Mar 09 2015](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-77936363)

Okay, I think you have a good grasp on it, and I trust what you decide :)

---

@IagoSRL commented on [Tue Mar 10 2015](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-78051997)

Josh, I decided already, just posted the final details and done. Just sometimes feels good to share thinking, comment on things to clarify ideas, you know, sorry if lot of technique details.

---

@joshdanielson commented on [Tue Mar 10 2015](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-78088176)

I want to learn the technical details (as much as I can) so I appreciate you sharing them!  Just don't let me make decisions that you don't think I fully understand :)

---

@IagoSRL commented on [Tue Mar 10 2015](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-78129041)

Added the autocomplete feature for the /job-titles API.

---

@IagoSRL commented on [Thu May 28 2015](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-106305226)

From some weeks ago, the REST API for Availability v1.0 is complete, I just updated the status and information related to v1.1 needs.
**But now**, with the need for "REST API support to create Client Location by Freelancer: #725" is open to be the last remaining thing if included in this release (I added it to the task list in a new section 'DUDE').

Added the commented change on the Availability weekly schedule like for v1.1, per discussion on the UI for both App and Web (that will share replacing the current web UI, with multiple time ranges per day).

---

@joshdanielson commented on [Thu May 28 2015](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-106528501)

Great work! I'll use this list when I speak to Noah.

---

@IagoSRL commented on [Thu Aug 27 2015](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-135509155)

For client-booking, some more APIs will be needed needed by other issues (scheduled later), and some other needed now not added still like:
- /client-booking/calculate Server side calculating for selected pricing, taking care of freelancer preferences, bookCode (book me now button) and relationship that means some booking-type and fees or another, added gratuity if any, promotional code discounts (future). For the APP, will be a bit more slow to request server to get calculation, but may simplify the needed APIs that need to expose things like how to pick a booking-type or another from the App side, request fees for that and calculate them in a consistent way with the server.

---

@joshdanielson commented on [Wed Sep 09 2015](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-139001059)

Saw your updates but noticed ProviderClient isn't ServiceProfessionalClient. Not sure if it was overlooked or intentional! 

---

@IagoSRL commented on [Wed Oct 07 2015](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-146184234)

Hi Josh, I have the changes on the availability API done (to manage conversion of date-time without problems between server and client, sent in UTC), it has a complete internal rewrite of the CA2S code for computing availability, that means the API returns now a more concise output (can be even better, some notes added for some small further optimizations) and its done in **lot less time**.

To test it, go to the test REST page: http://dev.loconomics.com/en-US/tests/testrest
Locate the links for 'User Availability', the first one '/times' is the new, the '/slots' and '/dates' are the previous ones.
Click on the first link, go up and press the 'Get' button to try. Look how much time takes; try twice. Now do the same with the second and third links. Do you see the performance difference? :-)

---

@IagoSRL commented on [Wed Oct 07 2015](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-146184611)

Well, it's done the new code to compute availability, and connected through the new API exposed, but some internals still use the previous and I need to update them taking care of some details, but that's was the hard part.

---

@joshdanielson commented on [Wed Oct 07 2015](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-146265486)

AMAZING difference (I tried all three) :-) Great work, and I'm glad you took the extra time.

Looking forward to chatting tomorrow. Lots of good stuff happening on the marketing of the event and hoping for a good meeting with the new programmer guy tonight. 

---

@IagoSRL commented on [Wed Oct 07 2015](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-146268216)

THANKS!   I'm happy, I expected good results based on the analysis and the
my idea for a new approach, but on the way I had moments thinking that was
not able to complete it...but I was :))))))

And now, if some new details happens in next days or future, I will be
better prepared to do changes on this logic (and is more commented than
previously so good for collaborators too I hope) and to run some unit
testing to ensure keeps working properly.

El mié., 7 oct. 2015 a las 19:13, Joshua Danielson (<
notifications@github.com>) escribió:

> AMAZING difference (I tried all three) :-) Great work, and I'm glad you
> took the extra time.
> 
> Looking forward to chatting tomorrow. Lots of good stuff happening on the
> marketing of the event and hoping for a good meeting with the new
> programmer guy tonight.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/dani0198/Loconomics/issues/530#issuecomment-146265486
> .
> 
> ## 
> 
> Iago Lorenzo Salgueiro

---

@IagoSRL commented on [Thu Nov 26 2015](https://github.com/joshdanielson/loconomics-internal/issues/530#issuecomment-159929669)

@dani0198 Moved this issue to Sandbox, since is more than just one release.
I keep for myself a pinned tab with this issue so I can check any documentation and sub-task I added here, and I try to keep it updated. Will close when all tasks done or better moved specific tasks to its own issue and release.
Let me manage it :-)
